drv/ekf/ekf.c

```diff
@@ -0,0 +1,481 @@
+    // ##################################################
+    //
+    // Calculation performed is the following
+    // y = z - h(x)
+    //
+    // Observation model is:
+    //
+    // H =  [1, 0, 0, 0, 0]
+    //      [0, 1, 0, 0, 0]
+    //
+
+    y[0] = z[0] - ekf_p->x;
+    y[1] = z[1] - ekf_p->y;
+
+    // ##################################################
+    //                  Update State
+    // ##################################################
+    //
+    // Calculations is:
+    // x = x + K @ y
+    ekf_p->x     += K[0] * y[0] + K[1] * y[1];
+    ekf_p->y     += K[2] * y[0] + K[3] * y[1];
+    ekf_p->theta += K[4] * y[0] + K[5] * y[1];
+    ekf_p->V     += K[6] * y[0] + K[7] * y[1];
+    ekf_p->W     += K[8] * y[0] + K[9] * y[1];
+
+    // Constraint theta angle to [-pi, +pi]
+    ekf_p->theta = _EKF_angle_overflow(ekf_p->theta);
+
+    // ##################################################
+    //              Update Posterior P
+    // ##################################################
+    _EKF_update_post_P_lh2_xy(ekf_p, K);
+
+}
+
+void EKF_update_gyro_W(KalmanFilter_t *ekf_p, float z)
+{
+    // Define vector variables
+    float y;  // residual (1,1),  y = z - h(x)
+    float K[5]; // Kalman gain (5,1), K = P H.t (H P H.t + R)
+
+    // ##################################################
+    //             Calculate the Kalman Gain
+    // ##################################################
+    //
+    // Calculation performed is the following
+    // K = P H.t (H P H.t + R)
+    //
+    _EKF_compute_K_gyro_W(ekf_p, K);
+
+    // ##################################################
+    //              Calculate the Residual
+    // ##################################################
+    //
+    // Calculation performed is the following
+    // y = z - h(x)
+    //
+    // Observation model is:
+    //
+    // H =  [0, 0, 0, 0, 1]
+    //
+    y = z - ekf_p->W;
+
+
+    // ##################################################
+    //                  Update State
+    // ##################################################
+    //
+    // Calculations is:
+    // x = x + K @ y
+    ekf_p->x += K[0] * y;
+    ekf_p->y += K[1] * y;
+    ekf_p->theta += K[2] * y;
+    ekf_p->V += K[3] * y;
+    ekf_p->W += K[4] * y;
+
+    // Constraint theta angle to [-pi, +pi]
+    ekf_p->theta = _EKF_angle_overflow(ekf_p->theta);
+
+    // ##################################################
+    //              Update Posterior P
+    // ##################################################
+    _EKF_update_post_P_gyro_W(ekf_p, K);
+
+}
+
+void EKF_set_state(KalmanFilter_t *ekf_p,float *x0)
+{
+
+    // Copy inital state x0
+    ekf_p->x     = x0[0];
+    ekf_p->y     = x0[1];
+    ekf_p->theta = x0[2];
+    ekf_p->V     = x0[3];
+    ekf_p->W     = x0[4];
+}
+
+void EKF_set_P(KalmanFilter_t *ekf_p,float *P0)
+{
+
+    // Copy the initial Covariance
+    memcpy(ekf_p->P, P0, 25*4); // 25*4 becasue we are copying 25 floats.
```

Comment:
```suggestion
    memcpy(ekf_p->P, P0, 25 * sizeof(float)); // 25*4 because we are copying 25 floats.
```

And maybe use a define for 25 since it's used all other the place